src/QsCompiler/QirGeneration/Interop.cs

```diff
@@ -0,0 +1,436 @@
+        /// </summary>
+        /// <typeparam name="TIn">The type of the items in the given sequence.</typeparam>
+        /// <typeparam name="TOut">The type of the items in the returned array.</typeparam>
+        /// <param name="map">The function to apply to each item in the sequence.</param>
+        /// <param name="items">The sequence of items to map.</param>
+        private static TOut[] WithoutNullValues<TIn, TOut>(Func<TIn, TOut?> map, IEnumerable<TIn> items)
+            where TOut : class =>
+            items.Select(map).Where(i => i != null).Select(i => i!).ToArray();
+
+        /// <summary>
+        /// Bitcasts the given value to the expected type if needed.
+        /// Does nothing if the native type of the value already matches the expected type.
+        /// </summary>
+        private Value CastToType(Value value, ITypeRef expectedType) =>
+            value.NativeType.Equals(expectedType)
+            ? value
+            : this.sharedState.CurrentBuilder.BitCast(value, expectedType);
+
+        /// <inheritdoc cref="MapToInteropType(ITypeRef)"/>
+        private ITypeRef? MapToInteropType(ResolvedType type) =>
+            this.MapToInteropType(this.sharedState.LlvmTypeFromQsharpType(type));
+
+        /// <summary>
+        /// Maps the given Q#/QIR type to a more interop-friendly type.
+        /// Returns null only if the given type is Unit. Strips items of type Unit inside tuples.
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// The given type is a pointer to a non-struct type,
+        /// or the given type is not specified in the QIR format.
+        /// </exception>
+        private ITypeRef? MapToInteropType(ITypeRef t)
+        {
+            // Range, Tuple (typed and untyped), Array, Result, String, BigInt, Callable, and Qubit
+            // are all structs or struct pointers.
+            t = t.IsPointer ? Types.StructFromPointer(t) : t;
+            var typeName = (t as IStructType)?.Name;
+
+            var bytePtrType = this.sharedState.Context.Int8Type.CreatePointerType();
+
+            if (typeName == TypeNames.Array || typeName == TypeNames.BigInt)
+            {
+                return this.sharedState.Context.CreateStructType(
+                    packed: false,
+                    this.sharedState.Context.Int64Type,
+                    this.sharedState.Types.DataArrayPointer)
+                   .CreatePointerType();
+            }
+            else if (typeName == TypeNames.String)
+            {
+                return this.sharedState.Types.DataArrayPointer;
+            }
+            else if (typeName == TypeNames.Callable || typeName == TypeNames.Qubit)
+            {
+                return bytePtrType;
+            }
+            else if (typeName == TypeNames.Result)
+            {
+                return this.sharedState.Context.Int8Type;
+            }
+            else if (t is IStructType st)
```

Comment:
Does `Range` fall into the `IStructType` category?

Currently the way an entry point `Range` argument is generated as `%Range`. However, since the size of `%Range` is 24 bytes, interop with C++ is a bit strange because QIR expects an argument of that size but from C++ a 24 bytes single argument cannot be passed so three 8 bytes arguments have to be used.

For interop purposes, I think a `Range` argument in the entry-point operation should be received as a pointer `%Range*`.

Reply:
I'll provide a bit more details to Cesar's observation.
During [that discussion](https://github.com/microsoft/qsharp-runtime/pull/520#discussion_r589733387), I asked Cesar offline to stop in the debugger at the caller and see in the disassembly how the caller passes the instance of the `struct Range` by value.
From Cesar's analysis I got an impression that the caller creates a copy (of the `struct Range` instance) on the stack (that's what I expected), but then passes to the callee the _address_ of that copy. As a result the callee, instead of getting the three 64-bit values in certain part of the stack, gets (in that part of the stack or in a CPU register) the _address_ to those values.

Reply:
Indeed there is an issue tracked on this already: https://github.com/microsoft/qsharp-compiler/issues/801
I believe the issue only occurs for interop, and indeed Range would fall under this IStructType here and should be passed as a pointer. However, I think I'll have to update the PR to make sure the actual memory storing the struct is allocated on the heap such that it won't be deallocated upon exiting QIR (I forgot).

Reply:
The ranges should now be handled properly (confirmation pending for whether everything executes as it should).